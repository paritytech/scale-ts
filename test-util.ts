/// <reference lib="deno.unstable"/>

import { assertEquals, assertThrows } from "https://deno.land/std@0.161.0/testing/asserts.ts"
import { assertSnapshot } from "https://deno.land/std@0.161.0/testing/snapshot.ts"
import { AnyShape, assert, DecodeBuffer, Shape, ShapeAssertError } from "./common/mod.ts"

const [lipsum, words, cargoLock] = ["lipsum.txt", "words.txt", "Cargo.lock"].map((fileName) => () =>
  Deno.readTextFile(fileName)
)
export const files = { lipsum: lipsum!, words: words!, cargoLock: cargoLock! }

export function testShape<T>(
  shape: Shape<T>,
  values: NoInfer<T>[] | Record<string, NoInfer<T> | (() => NoInfer<T>)>,
  async?: boolean,
): void
export function testShape<T>(
  shape: Shape<T>,
  values: T[] | Record<string, T | (() => T)>,
  async = false,
) {
  for (const key in values) {
    let value = values[key as never] as T | (() => T)
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => T)()
      }
      assert(shape, value)
      const encoded = async ? await shape.encodeAsync(value) : shape.encode(value)
      await assertSnapshot(t, encoded, { serializer: serializeU8A })
      const decodeBuffer = new DecodeBuffer(encoded)
      const decoded = shape.subDecode(decodeBuffer)
      assertEquals(decoded, value)
      assertEquals(decodeBuffer.index, encoded.length)
      assert(shape, decoded)
    })
  }
}

export function testInvalid(shape: AnyShape, values: unknown[] | Record<string, unknown | (() => unknown)>) {
  for (const key in values) {
    let value: unknown = values[key as never]
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} invalid ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => unknown)()
      }
      await assertThrowsSnapshot(t, () => assert(shape as Shape<any>, value), ShapeAssertError)
    })
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n")
}

type NoInfer<T> = T extends infer U ? U : never
export function benchShape<T>(name: string, shape: Shape<T>, value: NoInfer<T>): void
export function benchShape<T>(name: string, shape: Shape<T>, value: T) {
  const encoded = shape.encode(value)
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    shape.encode(value)
  })
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    shape.decode(encoded)
  })
}

export async function assertThrowsSnapshot(
  t: Deno.TestContext,
  fn: () => unknown,
  ctor: abstract new(...args: any) => Error = Error,
) {
  try {
    fn()
  } catch (error) {
    if (!(error instanceof ctor)) {
      throw new Error("Expected " + ctor.name)
    }
    return await assertSnapshot(t, `${error.name}: ${error.message}`, { serializer: (x) => x })
  }
  throw new Error("Expected function to throw")
}

export { assertEquals, assertSnapshot, assertThrows }
