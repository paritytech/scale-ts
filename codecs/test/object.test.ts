import * as $ from "../../mod.ts";
import { testCodec } from "../../test-util.ts";

const $person = $.withMetadata(
  $.metadata("$person"),
  $.object(
    ["name", $.str],
    ["nickName", $.str],
    ["superPower", $.option($.str)],
    ["luckyNumber", $.u8],
  ),
);

testCodec($person, [
  {
    name: "Darrel",
    nickName: "The Durst",
    superPower: "telekinesis",
    luckyNumber: 9,
  },
  {
    name: "Michael",
    nickName: "Mike",
    superPower: undefined,
    luckyNumber: 7, // Cummon... be more predictable!
  },
]);
