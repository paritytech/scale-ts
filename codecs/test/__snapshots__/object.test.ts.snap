export const snapshot = {};

snapshot[`\$person { name: "Darrel", nickName: "The Durst", superPower: "telekinesis", luckyNumber: 9 } 1`] = `
18
44
61
72
72
65
6c
24
54
68
65
20
44
75
72
73
74
01
2c
74
65
6c
65
6b
69
6e
65
73
69
73
09
`;

snapshot[`\$person { name: "Michael", nickName: "Mike", luckyNumber: 7 } 1`] = `
1c
4d
69
63
68
61
65
6c
10
4d
69
6b
65
00
07
`;

snapshot[`\$person invalid null 1`] = `ScaleAssertError: value == null`;

snapshot[`\$person invalid undefined 1`] = `ScaleAssertError: typeof value !== "object"`;

snapshot[`\$person invalid 123 1`] = `ScaleAssertError: typeof value !== "object"`;

snapshot[`\$person invalid [Function (anonymous)] 1`] = `ScaleAssertError: typeof value !== "object"`;

snapshot[`\$person invalid {} 1`] = `ScaleAssertError: !("name" in value)`;

snapshot[`\$person invalid { name: "", nickName: "", superPower: 0, luckyNumber: 0 } 1`] = `ScaleAssertError: typeof value.superPower !== "string"`;

snapshot[`\$person invalid { name: "", nickName: "", superPower: "", luckyNumber: -1 } 1`] = `ScaleAssertError: value.luckyNumber < 0`;

snapshot[`\$person invalid { name: "", nickName: "", superPower: "", unluckyNumber: 0 } 1`] = `ScaleAssertError: !("luckyNumber" in value)`;

snapshot[`\$.object(
  \$.taggedUnion(
    "_tag",
    [
      Variant { tag: "a", codec: \$.object(Array) },
      Variant { tag: "b", codec: \$.object(Array) }
    ]
  ),
  \$.field("bar", \$.u8)
) { _tag: "a", bar: 123 } 1`] = `
00
7b
`;

snapshot[`\$.object(
  \$.taggedUnion(
    "_tag",
    [
      Variant { tag: "a", codec: \$.object(Array) },
      Variant { tag: "b", codec: \$.object(Array) }
    ]
  ),
  \$.field("bar", \$.u8)
) { _tag: "b", x: 0, bar: 123 } 1`] = `
01
00
7b
`;

snapshot[`\$.object(
  \$.taggedUnion(
    "_tag",
    [
      Variant { tag: "a", codec: \$.object(Array) },
      Variant { tag: "b", codec: \$.object(Array) }
    ]
  ),
  \$.field("bar", \$.u8)
) invalid null 1`] = `ScaleAssertError: value == null`;

snapshot[`\$.object(
  \$.taggedUnion(
    "_tag",
    [
      Variant { tag: "a", codec: \$.object(Array) },
      Variant { tag: "b", codec: \$.object(Array) }
    ]
  ),
  \$.field("bar", \$.u8)
) invalid { _tag: null, bar: 1 } 1`] = `ScaleAssertError: typeof value._tag !== "string"`;

snapshot[`\$.object(
  \$.taggedUnion(
    "_tag",
    [
      Variant { tag: "a", codec: \$.object(Array) },
      Variant { tag: "b", codec: \$.object(Array) }
    ]
  ),
  \$.field("bar", \$.u8)
) invalid { _tag: "", bar: 1 } 1`] = `ScaleAssertError: value._tag: invalid tag`;

snapshot[`\$.object(
  \$.taggedUnion(
    "_tag",
    [
      Variant { tag: "a", codec: \$.object(Array) },
      Variant { tag: "b", codec: \$.object(Array) }
    ]
  ),
  \$.field("bar", \$.u8)
) invalid { _tag: "b", bar: 1 } 1`] = `ScaleAssertError: !("x" in value)`;

snapshot[`\$.object(
  \$.taggedUnion(
    "_tag",
    [
      Variant { tag: "a", codec: \$.object(Array) },
      Variant { tag: "b", codec: \$.object(Array) }
    ]
  ),
  \$.field("bar", \$.u8)
) invalid { _tag: "b", x: -1, bar: 1 } 1`] = `ScaleAssertError: value.x < 0`;

snapshot[`\$.object(
  \$.taggedUnion(
    "_tag",
    [
      Variant { tag: "a", codec: \$.object(Array) },
      Variant { tag: "b", codec: \$.object(Array) }
    ]
  ),
  \$.field("bar", \$.u8)
) invalid { _tag: "a", bar: -1 } 1`] = `ScaleAssertError: value.bar < 0`;
